### PR TITLE
ENH: AlphaGo to MuZero

### PR DESCRIPTION
**Pull request by AntoineTheb**
This is a long one and might require an entire lab meeting

### CODE REVIEW

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+---
+layout: review
+title: "AlphaGo/AlphaGoZero/AlphaZero/MuZero: Mastering games using progressively fewer priors"
+tags: deep-learning reinforcement
+author: "Antoine Théberge"
+cite:
+    authors: "Julian Schrittwieser, Ioannis Antonoglou, Thomas Hubert, Karen Simonyan, Laurent Sifre, Simon Schmitt, Arthur Guez, Edward Lockhart, Demis Hassabis, Thore Graepel, Timothy Lillicrap, David Silver"
+    title:   "Mastering Atari, Go, Chess and Shogi by Planning with a Learned Model"
+    venue:   "arXiv"
+pdf: ""
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:10]: Could you please add the link to the paper?

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+## Monte Carlo Tree Search
+
+![](/article/images/alpha/mcts.jpg)
+> source: wikipedia.org
+
+Monte Carlo tree search (MCTS) was actually first introduced in 2006 to play the game of Go[^1] and won the 10th computer-Go tournament. It has since then been applied to many problems including chess, shogi, bridge, poker, and video games such as Starcraft.
+
+As the name implies, MCTS is a tree search algorithm that uses Monte Carlo sampling of the search space to iteratively refine the evaluation of nodes. From a root node, child nodes are traversed until a leaf node (a node that has one or more children that has not been evaluated or until a maximum depth is reached) is reached. Then, a child is selected and a rollout (selecting actions according to a possibly random policy until the end of the game) is performed, assigning the result to the node. Finally, the node's value is "backpropagated" to parent nodes. Finally, the optimal action can be inferred from the node's values. This process is repeated for each move.
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:30]: Not sure if this sentence requires being re-worded: "From a root node, child nodes are traversed until a leaf node (a node that has one or more children that has not been evaluated or until a maximum depth is reached) is reached."

Can a leaf node have children?

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+$$a_t = \argmax_a U(s,a)$$
+
+with
+
+$$U(s,a) = Q(s,a) + c \cdot p(s,a) \cdot \frac{\sqrt{\sum_b N(s,b)}}{1+N(s,a)}$$
+
+$$Q(s,a) = \frac{1}{N(s,a)}\sum_{s'|s,a\rightarrow s'}v(s')\; \text{(an average of the values of the nodes from s' to s)}$$ 
+
+and $$c$$ a parameter promoting exploration. However, instead of performing rollouts, the network uses $$v(s_L)$$ to predict the outcome. MCTS' execution gives the tuple $$(s, \pi, p, \_)$$ where $$s$$ is the game's state, $$\pi$$ corresponds to the normalized state/actions visitations
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:107]: Not sure whether the term "visitation" is appropriate in this context.

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+$$\pi = \frac{N(s, \cdot)^{1/\tau}}{\sum_b N(s,b)^{1/\tau}}$$
+
+with $$\tau$$ a temperature parameter controlling exploration. $$p$$ is the action probabilities for state $$s$$, and $$\_$$ is later filled with $$z \in \pm 1$$ depending on the game's outcome. Then, the loss function
+
+$$l = (z - v)^2 - \pi^T \log p + c \|\theta\|^2$$
+
+where $$c$$ in this case is a regularization term, is used to train the network. We can interpret the loss as cross-entropy forcing the policy to match MCTS' behavior.
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:115]: (...) is a regularization term used to train the network. We can interpret the loss as a cross-entropy

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+![](/article/images/alpha/fig5.jpg)
+> Silver, D., Schrittwieser, J., Simonyan, K. et al. Mastering the game of Go without human knowledge. Nature 550, 354–359 (2017). https://doi.org/10.1038/nature24270
+
+From the above figure, we can see that the algorithm can surpass AlphaGo after only ~40h of training. Interestingly enough, we can see that AlphaGo Zero is not really good at prediction moves, but quite good at predicting game outcome. A supervised learning version of the algorithm is shown for comparison. Because it manages to outperform AlphaGo (and therefore the best human player is ELO is transitive), the results seem to imply that AlphaGo Zero learns to play differently than human players (which could explain moves 37 and 78).
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:120]: (...) at move prediction
or
(...) at predicting moves
?

Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:120]: (and therefore the best human player is ELO is transitive)
requires re-wording maybe?

File: article/_posts/2020-11-19-AlphaMuGoZero.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+![](/article/images/alpha/fig6.jpg)
+![](/article/images/alpha/tab5.jpg)
+> Silver, David, et al. “Mastering Chess and Shogi by Self-Play with a General Reinforcement Learning Algorithm.” ArXiv:1712.01815 [Cs], Dec. 2017. arXiv.org, http://arxiv.org/abs/1712.01815
+
+Above are shown results from the training process. We can observe that AlphaZero manages to defeat both AlphaGo Zero as well as game-specific algorithms.
+
+## MuZero
+
+Chess, Shogi and Go are games that can easily offer a perfect simulator as their rules are simple and well defined. However, this is not the case for most real-world scenarios such as robotics. Model-based reinforcement learning tries to tackle this problem by learning a simulator and then allowing planning using this learned simulator.
```
Comment (jhlegarreta) [inline article/_posts/2020-11-19-AlphaMuGoZero.md:143]: Since the example given is about Atari games and not robotics, may be this sentence
(...) most real-world scenarios such as robotics.
can be re-worded as
(...) most real-world scenarios, such as robotics or other continuous-space games and problems.

**Review by jhlegarreta — COMMENTED** [review]
As always, interesting stuff.

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 1 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Done in 137ef52 !

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 3 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Reworded in 137ef52

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 4 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Fixed in 137ef52

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 2 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Reworded in 137ef52

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 5 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Fixed in 137ef52

Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: woops I a word there. Fixed in 137ef52

File: article/_posts/2020-11-19-AlphaMuGoZero.md (hunk 6 above)
Comment (AntoineTheb) [reply article/_posts/2020-11-19-AlphaMuGoZero.md]: Fixed ! 137ef52

**Review by jhlegarreta — APPROVED** [review]
:+1:

**Review by jodoin — APPROVED** [review]

**Review by jhlegarreta — APPROVED** [review]
> I approved the PR, but then realized that there are a few screenshots from at least a Nature paper, and I am not sure that it is open access.

Talked to Antoine. The *Nature* paper is open access.
